src/iqm/qiskit_iqm/iqm_provider.py

```diff
@@ -241,119 +241,148 @@ def serialize_circuit(self, circuit: QuantumCircuit) -> Circuit:
-            elif instruction.name == 'rx':
-                angle_t = float(instruction.params[0] / (2 * np.pi))
-                native_inst = Instruction(name='prx', qubits=qubit_names, args={'angle_t': angle_t, 'phase_t': 0.0})
-            elif instruction.name == 'y':
-                native_inst = Instruction(name='prx', qubits=qubit_names, args={'angle_t': 0.5, 'phase_t': 0.25})
-            elif instruction.name == 'ry':
-                angle_t = float(instruction.params[0] / (2 * np.pi))
-                native_inst = Instruction(name='prx', qubits=qubit_names, args={'angle_t': angle_t, 'phase_t': 0.25})
-            elif instruction.name == 'cz':
-                native_inst = Instruction(name='cz', qubits=qubit_names, args={})
-            elif instruction.name == 'move':
-                native_inst = Instruction(name='move', qubits=qubit_names, args={})
-            elif instruction.name == 'barrier':
-                native_inst = Instruction(name='barrier', qubits=qubit_names, args={})
-            elif instruction.name == 'measure':
-                if len(circuit_instruction.clbits) != 1:
-                    raise ValueError(
-                        f'Unexpected: measurement instruction {circuit_instruction} uses multiple classical bits.'
-                    )
-                clbit = circuit_instruction.clbits[0]  # always a single-qubit measurement
-                mk = str(MeasurementKey.from_clbit(clbit, circuit))
-                native_inst = Instruction(name='measure', qubits=qubit_names, args={'key': mk})
-                clbit_to_measure[clbit] = native_inst
-            elif instruction.name == 'reset':
-                # implemented using a measure instruction to measure the qubits, and
-                # one cc_prx per qubit to conditionally flip it to |0>
-                feedback_key = '_reset'
+        instructions = serialize_instructions(circuit, self._idx_to_qb, ignore_nonnative_gates=False)
+
+        try:
+            metadata = to_json_dict(circuit.metadata)
+        except ValueError:
+            warnings.warn(
+                f'Metadata of circuit {circuit.name} was dropped because it could not be serialised to JSON.',
+            )
+            metadata = None
+
+        return Circuit(name=circuit.name, instructions=instructions, metadata=metadata)
+
+
+def serialize_instructions(
+    circuit: QuantumCircuit, index_to_qubit_mapping: dict[int, str], ignore_nonnative_gates: bool
+) -> list[Instruction]:
+    """Serialize a quantum circuit into Instructions in the IQM data transfer format.
+
+    This is an internal helper for :meth:`.IQMBackend.serialize_circuit` that gives slightly more control.
+    See :meth:`.IQMBackend.serialize_circuit` for details.
+    There is usually no need to use this function directly.
+
+    Args:
+        circuit: quantum circuit to serialize
+        index_to_qubit_mapping: Mapping from logical to physical qubit names.
```

Comment:
```suggestion
        index_to_qubit_mapping: Mapping from qubit indices to the corresponding qubit names.
```
I think there are multiple use cases, some of which use logical and some of which use physical indices/names. Better to be imprecise.